### ISSUE TITLE
[rllib] Error with DDPG Trainer initialization: Input 'y' of 'Mul' Op has type float64 that does not match type float32 of argument 'x'

### ISSUE THREAD

**Issue by ZekiDorukErden** [ray-project/ray#9643]
Hello,

I encounter an error during the initialization of DDPG Trainer (I also tried PPO & A2C trainers and did not get the same error.) I understand it is about a type conversion, and full traceback goes as follows:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 470, in _apply_op_helper
    preferred_dtype=default_dtype)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 1317, in convert_to_tensor
    (dtype.name, value.dtype.name, value))
ValueError: Tensor conversion requested dtype float32 for Tensor with dtype float64: <tf.Tensor 'ppo_policy0/Select:0' shape=(1,) dtype=float64>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./run_simulation.py", line 528, in <module>
    "log_level": "ERROR",
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/agents/trainer_template.py", line 90, in __init__
    Trainer.__init__(self, config, env, logger_creator)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/agents/trainer.py", line 450, in __init__
    super().__init__(config, logger_creator)
  File "/usr/local/lib/python3.7/dist-packages/ray/tune/trainable.py", line 175, in __init__
    self._setup(copy.deepcopy(self.config))
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/agents/trainer.py", line 623, in _setup
    self._init(self.config, self.env_creator)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/agents/trainer_template.py", line 115, in _init
    self.config["num_workers"])
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/agents/trainer.py", line 696, in _make_workers
    logdir=self.logdir)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/evaluation/worker_set.py", line 59, in __init__
    RolloutWorker, env_creator, policy, 0, self._local_config)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/evaluation/worker_set.py", line 282, in _make_worker
    extra_python_environs=extra_python_environs)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/evaluation/rollout_worker.py", line 379, in __init__
    self._build_policy_map(policy_dict, policy_config)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/evaluation/rollout_worker.py", line 937, in _build_policy_map
    policy_map[name] = cls(obs_space, act_space, merged_conf)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/policy/tf_policy_template.py", line 143, in __init__
    obs_include_prev_action_reward=obs_include_prev_action_reward)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/policy/dynamic_tf_policy.py", line 222, in __init__
    explore=explore)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/utils/exploration/gaussian_noise.py", line 90, in get_exploration_action
    explore, timestep)
  File "/usr/local/lib/python3.7/dist-packages/ray/rllib/utils/exploration/ornstein_uhlenbeck_noise.py", line 102, in _get_tf_exploration_action_op
    noise = scale * self.ou_base_scale * ou_state_new * high_m_low
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/math_ops.py", line 984, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/math_ops.py", line 1283, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 6092, in mul
    "Mul", x=x, y=y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 506, in _apply_op_helper
    inferred_from[input_arg.type_attr]))
TypeError: Input 'y' of 'Mul' Op has type float64 that does not match type float32 of argument 'x'.
```

The trainer initialization is as follows:

```
trainer = DDPGTrainer(
              env=env_title,
              config={
                "num_workers": num_workers,
                "num_cpus_per_worker": num_cpus_per_worker,
                "num_gpus": num_gpus,
                "num_gpus_per_worker": num_gpus_per_worker,
                "model": nw_model,
                "lr": lr,
                "gamma": gamma,
                "actor_hiddens": [hidden_layer_size, hidden_layer_size],
                "critic_hiddens": [hidden_layer_size, hidden_layer_size],
                "multiagent": {
                  "policy_graphs": policy_graphs,
                  "policy_mapping_fn": policy_mapping_fn,
                  "policies_to_train": ["ppo_policy{}".format(i) for i in range(n_agents)],
                },
                "callbacks": {
                  "on_episode_start": tune.function(on_episode_start),
                  "on_episode_step": tune.function(on_episode_step),
                  "on_episode_end": tune.function(on_episode_end),
                },
                "log_level": "ERROR",
              })
```

Furthermore, I did not get the same error when I am running the code in CoLab.

There is no place in my script where float64 or float32 is specified. Any ideas on what may be the case?

System Information:
Python version 3.7
Tensorflow version 2.2.0
Ray version 0.8.6
OS: Ubuntu 16.04

**Comment by duburcqa** [comment]
> There is no place in my script where float64 or float32 is specified. Any ideas on what may be the case?

Numpy defaults to float32, while Torch defaults to float64. So converting a Numpy array in Torch tensor has different dtype than "native" torch tensor, and conversely.

**Comment by HJ-TANG** [comment]
I have the same issue now. but I didn't get the error when I run the code a couple of weeks ago.

**Comment by alexanderkell** [comment]
I am getting the same error when running with Ubuntu. I don't get this issue when running on macOS

**Comment by alexanderkell** [comment]
I managed to fix this by downgrading ray to 0.8.4 with pip3 install ray==0.8.4

**Comment by alexanderkell** [comment]
Has any reason for this been found? The hack I posted in September still works, but I would like to use the latest version of Ray.
